[PyTorch] Replace IValue::intrusive_ptr with IValue::isIntrusivePtr()

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76425
* #76418

This tidies up an extra piece of state that was computable from the other state in IValue.

Differential Revision: [D31290017](https://our.internmc.facebook.com/intern/diff/D31290017/)

swolchok committed:
[PyTorch] Replace IValue::intrusive_ptr with IValue::isIntrusivePtr()

This tidies up an extra piece of state that was computable from the other state in IValue.

Differential Revision: [D31290017](https://our.internmc.facebook.com/intern/diff/D31290017/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Replace IValue::intrusive_ptr with IValue::isIntrusivePtr()"


This tidies up an extra piece of state that was computable from the other state in IValue.

Differential Revision: [D31290017](https://our.internmc.facebook.com/intern/diff/D31290017/)

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -365,11 +365,14 @@ struct TORCH_API IValue final {
     }
 
     if (payload.u.as_intrusive_ptr == c10::UndefinedTensorImpl::singleton()) {
+      if (isStorage() || isGenerator()) {
+        return 1;
+      }
```
This deserves a comment.

swolchok replied:
changing to use isIntrusivePtrLegacyBehavior

peterbell10 commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -921,7 +913,13 @@ struct TORCH_API IValue final {
+    if (isTensor()) {
+      return payload.as_tensor.defined();
+    }
+    if (isStorage() || isGenerator()) {
+      return payload.u.as_intrusive_ptr != c10::UndefinedTensorImpl::singleton();
+    }
+    return isIntrusivePtr();
```
I presume this is for BC, so wouldn't it make sense to use `isIntrusivePtrLegacyBehavior()`?

swolchok replied:
nice catch!

swolchok committed:
Update on "[PyTorch] Replace IValue::intrusive_ptr with IValue::isIntrusivePtr()"


This tidies up an extra piece of state that was computable from the other state in IValue.

Differential Revision: [D31290017](https://our.internmc.facebook.com/intern/diff/D31290017/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Replace IValue::intrusive_ptr with IValue::isIntrusivePtr()"


This tidies up an extra piece of state that was computable from the other state in IValue.

Differential Revision: [D31290017](https://our.internmc.facebook.com/intern/diff/D31290017/)

[ghstack-poisoned]

ezyang commented:
@JackCaoG XLA may need to track this change

swolchok commented:
not a believer on the two CI failures

@pytorchbot merge this please

swolchok commented:
@pytorchbot merge this please
